### ISSUE TITLE
Improve Documentation & Readability in and between LDA/HDP Implementations

### ISSUE THREAD

**Issue by benrbray** [RaRe-Technologies/gensim#262]
Gensim is arguably the most popular library for topic modeling (among other things), and as an open source project written in a clear, concise language, has the unique opportunity to provide practical insight into the structure of these models while continuing to maintain high standards of quality and efficiency.  

While learning about topic models I have referred to the gensim implementations countless times, as I am sure many others have.  It quickly becomes very obvious that gensim's implementations of `LdaMallet`, `LdaModel`, and `HDPModel` share a lot of common structure, and it may be enlightening to interested developers/researchers if the shared bits of code were merged into a common interface, as mentioned briefly in #223.

Merging common structure among implementations will also help with maintenance.  I suspect that many of the differences between `LdaModel` and `HDPModel` are the result of code from one being copied to write the other, which has since diverged as one has been more often modified.  Take, for example, the `dirichlet_expectation` function appearing in both [`LdaModel`](https://github.com/piskvorky/gensim/blob/develop/gensim/models/ldamodel.py#L56) and [`HDPModel`](https://github.com/piskvorky/gensim/blob/develop/gensim/models/hdpmodel.py#L71), which return the same result in two different ways, for no good reason.

I'm willing to spend the time making and proposing the necessary changes, as long as I'm sure such changes would be welcomed by the community.  This would involve:
- Refactoring bits of code similar among topic model implementations into a common interface
- Extracting code from some of the longer procedures into their own functions, for readability and comparison with the algorithm described in the corresponding paper, provided that there will be no added overhead
- Improving general documentation, variable names, and code organization for clarity and consistency

Does this sound like a desirable enhancement to the source?

Thanks!

**Comment by bhargavvader** [comment]
@tmylk does the new base class mean this can be closed?

**Comment by tmylk** [comment]
Once #996 is done, then this can be closed.

**Comment by menshikh-iv** [comment]
Resolved in #223, #1055, etc
